msgpack_unpack.c

```diff
@@ -694,66 +556,70 @@ int msgpack_unserialize_map_item(
         }
     }
 
-    if (Z_TYPE_PP(container) != IS_ARRAY && Z_TYPE_PP(container) != IS_OBJECT)
-    {
+    zend_bool container_is_ref = 0;
+    if (container != NULL && Z_ISREF_P(*container)) {
+        container_is_ref = 1;
+        ZVAL_DEREF(*container);
+    }
+
+    if (Z_TYPE_P(*container) != IS_ARRAY && Z_TYPE_P(*container) != IS_OBJECT) {
         array_init(*container);
     }
 
-    switch (Z_TYPE_P(key))
-    {
-        case IS_LONG:
-            if (zend_hash_index_update(
-                    HASH_OF(*container), Z_LVAL_P(key), &val,
-                    sizeof(val), NULL) == FAILURE)
-            {
-                zval_ptr_dtor(&val);
-                MSGPACK_WARNING(
-                    "[msgpack] (%s) illegal offset type, skip this decoding",
-                    __FUNCTION__);
-            }
-            zval_ptr_dtor(&key);
-            break;
-        case IS_STRING:
-            if (zend_symtable_update(
-                    HASH_OF(*container), Z_STRVAL_P(key), Z_STRLEN_P(key) + 1,
-                    &val, sizeof(val), NULL) == FAILURE)
-            {
-                zval_ptr_dtor(&val);
-                MSGPACK_WARNING(
-                    "[msgpack] (%s) illegal offset type, skip this decoding",
-                    __FUNCTION__);
-            }
-            zval_ptr_dtor(&key);
-            break;
-        default:
-            MSGPACK_WARNING("[msgpack] (%s) illegal key type", __FUNCTION__);
+    if (Z_TYPE_P(*container) == IS_OBJECT && Z_OBJCE_P(*container) != PHP_IC_ENTRY) {
```

Comment:
Calling zend_symtable_update on the HASH_OF a class breaks the class in PHP 7. Later when trying to modify the class you can't access members added via zend_symtable_update, and the class itself becomes immutable.

Is this expected behavior? Was msgpack-php relying on undefined behavior in <7 ?